src/nnet3/nnet-utils.cc

```diff
@@ -280,10 +261,31 @@ void SetLearningRate(BaseFloat learning_rate,
   }
 }
 
+void SetAsGradientNnet(Nnet *nnet) {
+  for (int32 c = 0; c < nnet->NumComponents(); c++) {
+    Component *comp = nnet->GetComponent(c);
+    if (comp->Properties() & kUpdatableComponent) {
+      UpdatableComponent *u_comp = dynamic_cast<UpdatableComponent*>(comp);
+      KALDI_ASSERT(u_comp != NULL);
+      u_comp->SetAsGradient();
+    }
+  }
+}
+
 void ScaleNnet(BaseFloat scale, Nnet *nnet) {
   if (scale == 1.0) return;
   else if (scale == 0.0) {
-    SetZero(false, nnet);
+    for (int32 c = 0; c < nnet->NumComponents(); c++) {
```

Comment:
I don't think we need to have a special case here any more, for scale == 0.0.
Just make sure that all the components that store stats [which do all support the
Scale() function], have a Scale() function that really sets everything to zero if
called with 0.0.